src/ILCompiler.Compiler/src/Compiler/InteropStubManager.cs

```diff
@@ -101,32 +103,165 @@ internal TypeDesc GetInlineArrayType(InlineArrayCandidate candidate)
+                        DelegateCreationThunk = InteropStateManager.GetForwardDelegateCreationThunk(delegateType)
                     };
             }
         }
 
-        internal HashSet<NativeStructType> GetStructMarshallingTypes()
+        internal struct StructMarshallingThunks
+        {
+            public TypeDesc StructType;
+            public NativeStructType NativeStructType;
+            public MethodDesc MarshallingThunk;
+            public MethodDesc UnmarshallingThunk;
+            public MethodDesc CleanupThunk;
+        }
+
+        internal IEnumerable<StructMarshallingThunks> GetStructMarshallingTypes()
+        {
+                foreach (var structType in _structMarshallingTypes)
+            {
+                yield return
+                    new StructMarshallingThunks()
+                    {
+                        StructType = structType,
+                        NativeStructType = InteropStateManager.GetStructMarshallingNativeType(structType),
+                        MarshallingThunk = InteropStateManager.GetStructMarshallingManagedToNativeThunk(structType),
+                        UnmarshallingThunk = InteropStateManager.GetStructMarshallingNativeToManagedThunk(structType),
+                        CleanupThunk = InteropStateManager.GetStructMarshallingCleanupThunk(structType)
+                    };
+            }
+        }
+
+        public void AddDependeciesDueToPInvoke(ref DependencyList dependencies, NodeFactory factory, MethodDesc method)
+        {
+            if (method.IsPInvoke)
+            {
+                dependencies = dependencies ?? new DependencyList();
+
+                MethodSignature methodSig = method.Signature;
+                AddDependenciesDueToPInvokeDelegate(ref dependencies, factory, methodSig.ReturnType);
+
+                for (int i = 0; i < methodSig.Length; i++)
+                {
+                    AddDependenciesDueToPInvokeDelegate(ref dependencies, factory, methodSig[i]);
+                }
+            }
+
+            if (method.HasInstantiation)
+            {
+                dependencies = dependencies ?? new DependencyList();
+                AddMarshalAPIsGenericDependencies(ref dependencies, factory, method);
+            }
+        }
+
+        public void AddInterestingInteropConstructedTypeDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
         {
-            return _structMarshallingTypes;
+            if (type.IsDelegate)
+            {
+                var delegateType = (MetadataType)type;
+                if (delegateType.HasCustomAttribute("System.Runtime.InteropServices", "UnmanagedFunctionPointerAttribute"))
+                {
+                    AddDependenciesDueToPInvokeDelegate(ref dependencies, factory, delegateType);
+                }
+            }
+        }
+
+        public void AddInterestingPInvokeStructDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
```

Comment:
Please delete this code. There is very little value keeping this empty method here. I don't think using StructLayout is the right thing to do (it'll include too many false positives).